Implementing solution for issue #1340

TobiasHT5 opened the pull request:
### Description of the problem

This PR is intended to implement a solution for issue #1340 which I raised previously. This PR adds the ability to create a custom tooltip to be displayed as a descriptive text for the column instead of just displaying the column name.

### Description of Changes

This PR introduces a new variable `tooltip` of type `kivy.properties.StringProperty` to contain the descriptive text for the column. The `column_data` property is also expanded to contain a tuple of 4 items instead of 3 as before. The items are arrange in the form `column_data = [ (column_name, column_size, column_tooltip, column_sort_function) ]`

### Screenshots of the solution

![image](https://user-images.githubusercontent.com/35004147/189613793-6c6793a0-ce14-45c8-a7f4-d8e917ddb9c5.png)

### Code for testing new changes

```python
from kivymd.app import MDApp
from kivymd.uix.datatables import MDDataTable
from kivy.metrics import dp

class TestApp(MDApp):
    def build (self):
        return MDDataTable(column_data = [("column 1",dp(30),"this is column one"), ("column2",dp(30),"this is column two")]

TestApp().run()
```

TobiasHT5 committed:
implementing solution for issue #1340

Issue #1340 suggests if it's possible to modify MDDataTable so that it's possible to add a custom text for the tooltip of te table in order to add more descriptive inormation about the table instead of just displaying table name. This is my attempt at implementing a solution for this issue

TobiasHT5 committed:
Merge branch 'kivymd:master' into master

HeaTTheatR commented:
@ht-thomas OK, thanks for the PR. But please be careful! There are syntax errors in your changes.

HeaTTheatR commented on kivymd/uix/datatables/datatables.py:
```diff
@@ -326,23 +334,32 @@ def __init__(self, **kwargs):
+                            text=col_heading[0],
+                            width=self.cols_minimum[i]
+                            table_data=self.table_data,
+                        )
```
![Снимок экрана 2022-09-12 в 15 53 43](https://user-images.githubusercontent.com/16930280/189658958-7c249fe2-1003-48e0-bbb2-552c3f0d0974.png)

TobiasHT5 replied:
Oh yeah, sorry about that. I had made changes in my version of kivymd and was trying to just hand type them into my contribution fork to ready them for a PR, guess I missed to throughly check before submitting. Let me fix that in case you haven't fixed it.

HeaTTheatR replied:
No, I didn't fix anything.

TobiasHT5 replied:
I fixed it, hope you can review the PR again

TobiasHT5 committed:
fix syntax error

added missing "," at the end of line 354

TobiasHT5 committed:
Merge branch 'master' of https://github.com/ht-thomas/KivyMD

HeaTTheatR commented on kivymd/uix/datatables/datatables.py:
```diff
@@ -351,7 +351,7 @@ def __init__(self, **kwargs):
                         if len(col_heading) == 3
                         else CellHandler(
```
You still haven't fixed the bug...

HeaTTheatR replied:
@ht-thomas 
![Снимок экрана 2022-09-12 в 16 38 59](https://user-images.githubusercontent.com/16930280/189669002-be4c4290-0d38-458f-82c5-66e30dbc8793.png)

TobiasHT5 committed:
fix syntax error

HeaTTheatR commented:
@ht-thomas Your changes contain bugs. Please, after applying the change, test the code with examples from the documentation...

TobiasHT5 committed:
fix syntax error

fix tootip to tooltip in line 362

TobiasHT5 commented:
I'm requesting for you to review my the PR again.

HeaTTheatR commented:
> I'm requesting for you to review my the PR again.

Your changes break the current API

TobiasHT5 commented:
Is it possible to work around it?

HeaTTheatR commented:
@ht-thomas It's just that your changes don't work with the very first example from the documentation for the `MDDataTable` widget.

TobiasHT5 commented:
I think it's because I introduced the tooltip argument to be the third argument where the sort function was supposed to be. What if I made it the last parameter? That would maintain the current API and also extend functionality.

HeaTTheatR commented:
@ht-thomas Any changes should not break the current API.

TobiasHT5 committed:
maintain exsiting API

modify column_data to accept a tuple of `(column_name,column_size,column_sort_function,column_tooltip)` to maintain the existing API and also extending it

TobiasHT5 commented:
![image](https://user-images.githubusercontent.com/35004147/190095176-ef7fe242-f661-4d4b-8987-60a46efe08e9.png)

I've refactored the code to maintain the existing API. Now one can write like this `column_data = [ (column_name, column_size, column_sort_function, column_tooltip) ]`

HeaTTheatR commented:
@ht-thomas 

Your recent changes don't work with your sample code:

```python
from kivymd.app import MDApp
from kivymd.uix.datatables import MDDataTable
from kivy.metrics import dp


class TestApp(MDApp):
    def build(self):
        return MDDataTable(
            column_data=[
                ("column 1", dp(30), "this is column one"),
                ("column2", dp(30), "this is column two"),
            ]
        )


TestApp().run()
```

TobiasHT5 commented:
The sample code is now nullified. We can use already the existing examples in the kivymd documentation.

HeaTTheatR commented:
@ht-thomas I don't understand. Then what is the essence of your changes? I don't see how to use it...

TobiasHT5 commented:
The New example will be like this,
```python
from kivymd.app import MDApp
from kivymd.uix.datatables import MDDataTable
from kivy.metrics import dp

class TestApp (MDApp):
    def build (self):
        return MDDataTable(column_data = [("Name",dp(30),None,"Name of client"), ("Employment",dp(30),lambda *sort_args:sort_function(*sort_args),"Employment status of client")]

TestApp().run()
```

A simple example usage could be like this. I've not written sort_function in the example, this is just an example of how to use it.

HeaTTheatR commented:
@ht-thomas OK I will merge this when I finish some started changes.

TobiasHT5 commented:
Alright, thank you @HeaTTheatR
